Fix debug impls section and other mistakes in the tutorial

antoniosarosi opened the pull request:
I was going through the [tutorial](https://salsa-rs.netlify.app/tutorial) and found some things that are outdated making some sections confusing. Particulary:

1. The last section of ["Defining the parser: memoized functions and inputs"](https://salsa-rs.netlify.app/tutorial/parser), named ["The `returns(ref)` annotation"](https://salsa-rs.netlify.app/tutorial/parser#the-returnsref-annotation), appears to be outdated because it starts with:

> You may have noticed that `parse_statements` is tagged with `#[salsa::tracked(returns(ref))]`

wich is not true. The definition of `parse_statements` in the example code is:

```rust
#[salsa::tracked]
pub fn parse_statements(db: &dyn crate::Db, source: SourceProgram) -> Program<'_>
```

2. The section [Defining the parser: debug impls and testing](https://salsa-rs.netlify.app/tutorial/debug#defining-the-parser-debug-impls-and-testing) talks about `DebugWithDb` which seems to have been removed in favor of `salsa::Database::attach`. I included an explanation of how that works instead.

Rest of fixes are typos.

antoniosarosi committed:
Fix debug impls section and typos in book

antoniosarosi committed:
Make examples runnable

antoniosarosi committed:
Remove `use` from code snippets, they still don't run

MichaReiser commented on book/src/tutorial/parser.md:
```diff
@@ -56,13 +56,3 @@ The **second** parameter to a tracked function is **always** some kind of Salsa
-So, when called, `parse_statements` will return an `&Vec<Statement>` rather than cloning the `Vec`.
-This is useful as a performance optimization.
-(You may recall the `returns(ref)` annotation from the [ir](./ir.md) section of the tutorial,
-where it was placed on struct fields, with roughly the same meaning.)
```
It would be nice if we move this block to the `ir` section and explain the `returns(ref)` (and deref etc) there

antoniosarosi replied:
I think it's better to explain it in the `parser.md` file because `ir` only talks about structs (input, tracked, interned) whereas `parser` section talks about tracked functions, I think adding `returns(ref)` there makes more sense.

antoniosarosi replied:
Yeah that could make sense. I added an explanation of `returns` for both functions and structs in the `parser` section, it could be split into `returns` for structs in `ir` and then `returns` for functions in `parser`.

MichaReiser reviewed (commented):
Thank you. This is great. I've one small suggestion

antoniosarosi committed:
Explain the `returns` attribute in functions and structs

antoniosarosi committed:
Add returns explanation for structs

antoniosarosi committed:
Explain `returns` for functions and add example

antoniosarosi commented:
@MichaReiser I added separate explanations for both `returns` on struct fields and `returns` on functions. I don't know if they are entirely accurate though.

I also added an example that made it easier for me to understand how the combinatorics of `returns` works when used in structs and functions. I can remove it if it doesn't make sense to have it there.

antoniosarosi committed:
Run `cargo fmt`

antoniosarosi committed:
Fix `clippy` lints

antoniosarosi committed:
Make `returns` sections shorter

MichaReiser commented:
The section should now have been updated
